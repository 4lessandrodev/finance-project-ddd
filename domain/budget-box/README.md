# BudgetBox - Aggregate

```json
{
    "id":"uuid",
    "ownerId":"uuid",
    "description": "valid_description", // Value Object
    "balanceAvaliable": 1000,
    "isPercentual": true,
    "budgetPercentage": 80, // Value Object
    "transactionsIds": ["uuid", "uuid"],
    "reasons": [
        { // Entity
            "id": "uuid"
            "description": "valid_descrip√≠ont"
        }
    ],
}
```

### Structure

- BudgetBox: Aggregate
- description: Value Object - Ok
- budgetPercentage: Value Object
- reason: Entity
