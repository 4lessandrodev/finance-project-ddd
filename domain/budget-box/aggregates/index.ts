export * from './budget-box/budget-box.aggregate';
