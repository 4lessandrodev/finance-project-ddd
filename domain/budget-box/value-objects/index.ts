export * from './description.value-object';
