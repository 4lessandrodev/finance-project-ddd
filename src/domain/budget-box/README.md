# BudgetBox - Aggregate

```json
{
  "id": "uuid",
  "ownerId": "uuid",
  "description": "valid_description",
  "balanceAvaliable": 1000,
  "isPercentual": true,
  "budgetPercentage": 80,
  "reasons": [
    {
      "id": "uuid",
      "description": "valid_descrip√≠ont"
    }
  ]
}
```

### Structure

- BudgetBox: Aggregate - Ok
- reason-description: Value Object - Ok
- budget-description: Value Object - Ok
- budget-id: Value Object - Ok
- budget-percentage: Value Object - Ok
- reason: Entity - Ok
- reason-id: Value Object - Ok
