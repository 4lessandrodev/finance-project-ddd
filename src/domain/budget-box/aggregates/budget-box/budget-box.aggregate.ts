import { AggregateRoot, Result, UniqueEntityID } from '@shared/index';
import {
  BudgetDescriptionValueObject,
  DEFAULT_BUDGET_PERCENTAGE_VALUE,
  BUDGET_PERCENTAGE_MAX_VALUE,
  PercentageValueObject,
  UserIdValueObject,
  ReasonDomainEntity,
} from '@domain/index';

export interface BudgetAggregateProps {
  ownerId: UserIdValueObject;
  description: BudgetDescriptionValueObject;
  balanceAvaliable: number;
  isPercentual: boolean;
  budgetPercentage: PercentageValueObject;
  reasons: ReasonDomainEntity[];
}
export class BudgetBoxAggregate extends AggregateRoot<BudgetAggregateProps> {
  private constructor(props: BudgetAggregateProps, id?: UniqueEntityID) {
    super(props, id);
  }

  get id(): UniqueEntityID {
    return this._id;
  }

  get ownerId(): UserIdValueObject {
    return this.props.ownerId;
  }

  get description(): BudgetDescriptionValueObject {
    return this.props.description;
  }

  get balanceAvaliable(): number {
    return this.props.balanceAvaliable;
  }

  get isPercentual(): boolean {
    return this.props.isPercentual;
  }

  get budgetPercentage(): PercentageValueObject {
    return this.props.budgetPercentage;
  }

  get reasons(): ReasonDomainEntity[] {
    return this.props.reasons;
  }

  public static create(
    props: BudgetAggregateProps,
    id?: UniqueEntityID,
  ): Result<BudgetBoxAggregate> {
    if (
      !props.isPercentual &&
      props.budgetPercentage.value !== BUDGET_PERCENTAGE_MAX_VALUE
    ) {
      props.budgetPercentage = PercentageValueObject.create(
        DEFAULT_BUDGET_PERCENTAGE_VALUE,
      ).getResult();
    }

    return Result.ok<BudgetBoxAggregate>(new BudgetBoxAggregate(props, id));
  }
}
