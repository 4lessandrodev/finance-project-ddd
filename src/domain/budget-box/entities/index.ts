export * from './reason.domain-entity';
