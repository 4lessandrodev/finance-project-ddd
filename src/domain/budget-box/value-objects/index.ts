export * from './budget-description/budget-description.value-object';
export * from './budget-id/budget-id.value-object';
export * from './percentage/percentage.value-object';
export * from './reason-description/reason-description.value-object';
export * from '../../shared/common/reason-id/reason-id.value-object';
