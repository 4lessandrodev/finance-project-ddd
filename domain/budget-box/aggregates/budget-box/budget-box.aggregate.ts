export interface BudgetAggregateProps {
  ownerId: UserIdValueObject;
  description: BudgetDescriptionValueObject;
  balanceAvaliable: number;
  isPercentual: boolean;
  budgetPercentage: PercentageValueObject;
  reasons: ReasonDomainEntity[];
}

import { AggregateRoot, Result, UniqueEntityID } from '../../../shared';
import { UserIdValueObject } from '../../../user/value-objects';
import { ReasonDomainEntity } from '../../entities';
import {
  BudgetDescriptionValueObject,
  PercentageValueObject,
} from '../../value-objects';

export class BudgetBoxAggregate extends AggregateRoot<BudgetAggregateProps> {
  private constructor(props: BudgetAggregateProps, id?: UniqueEntityID) {
    super(props, id);
  }

  get id(): UniqueEntityID {
    return this._id;
  }

  get ownerId(): UserIdValueObject {
    return this.props.ownerId;
  }

  get description(): BudgetDescriptionValueObject {
    return this.props.description;
  }

  get balanceAvaliable(): number {
    return this.props.balanceAvaliable;
  }

  get isPercentual(): boolean {
    return this.props.isPercentual;
  }

  get budgetPercentage(): PercentageValueObject {
    return this.props.budgetPercentage;
  }

  get reasons(): ReasonDomainEntity[] {
    return this.props.reasons;
  }

  public static create(
    props: BudgetAggregateProps,
    id?: UniqueEntityID,
  ): Result<BudgetBoxAggregate> {
    if (!props.isPercentual && props.budgetPercentage.value !== 100) {
      props.budgetPercentage = PercentageValueObject.create(100).getResult();
    }

    return Result.ok<BudgetBoxAggregate>(new BudgetBoxAggregate(props, id));
  }
}
