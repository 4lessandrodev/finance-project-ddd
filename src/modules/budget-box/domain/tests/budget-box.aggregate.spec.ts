import { DomainId } from 'types-ddd';
import BudgetDescriptionValueObject from '../budget-description.value-object';
import ReasonDescriptionValueObject from '../reason-description.value-object';
import { BudgetBoxMock } from './mock/budget-box.mock';
import { ReasonMock } from './mock/reason.mock';

describe('budget-box.aggregate', () => {

	const budgetBoxMock = new BudgetBoxMock();
	const reasonMock = new ReasonMock();

	it('should create a budget-box with 100% if it is not percentage', () => {

		const budgetBox = budgetBoxMock.domain({
			budgetPercentage: 20,
			isPercentage: false
		});

		expect(budgetBox.isSuccess).toBe(true);
		expect(budgetBox.getResult().budgetPercentage.value).toBe(100);
	});

	it('should create a valid budget-box aggregate with 100% if provide not percentage', () => {

		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: [reasonMock.model({ id: 'valid_id' })]
		});
		
		expect(budgetBox.getResult().toObject()).toMatchSnapshot();
		expect(budgetBox.isSuccess).toBe(true);
		expect(budgetBox.getResult().budgetPercentage.value).toBe(50);
	});

	it('should remove a reason from a budget box with success', () => {
		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: [reasonMock.model({ id: 'valid_id' })]
		}).getResult();

		const reasonId = DomainId.create('valid_id');

		const result = budgetBox.removeReasonById(reasonId);

		expect(result).toBeTruthy();
	});

	it('should do not remove a reason from a budget if not found', () => {
		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: [reasonMock.model({ id: 'valid_id' })]
		}).getResult();

		const reasonId = DomainId.create('invalid_id');

		const result = budgetBox.removeReasonById(reasonId);

		expect(result).toBeFalsy();
	});

	it('should add a new reason with success', () => {
		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: []
		}).getResult();

		const reason = reasonMock.domain({ id: 'valid_id' }).getResult();

		budgetBox.addReason(reason);

		expect(budgetBox.reasons).toHaveLength(1);
	});

	it('should get null on get reason by id and do not exists', () => {
		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: []
		}).getResult();

		const result = budgetBox.getReasonById(DomainId.create('invalid_id'));

		expect(result).toBeNull();
	});

	it('should get null on get reason by id and do not exists', () => {

		const reason = reasonMock.model({ id: 'valid_id' });

		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: [reason]
		}).getResult();

		const result = budgetBox.getReasonById(DomainId.create('valid_id'));

		expect(result?.toObject()).toEqual(reason);
	});

	it('should update reason description with success', () => {

		const reason = reasonMock.model({ id: 'valid_id', description: 'valid_description' });

		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: [reason]
		}).getResult();

		const ID = DomainId.create('valid_id');
		const description = ReasonDescriptionValueObject.create('updated_description').getResult();
		const result = budgetBox.changeReasonDescription(ID, description);

		expect(result).toBeTruthy();
		expect(budgetBox.getReasonById(ID)?.description.value).toBe('updated_description');
		
	});

	it('should return false if does not update description', () => {

		const budgetBox = budgetBoxMock.domain({
			isPercentage: true,
			budgetPercentage: 50,
			reasons: []
		}).getResult();

		const ID = DomainId.create('valid_id');
		const description = ReasonDescriptionValueObject.create('updated_description').getResult();
		const result = budgetBox.changeReasonDescription(ID, description);

		expect(result).toBeFalsy();
	});

	it('should change name with success', () => {

		const budgetBox = budgetBoxMock.domain({ description: 'old_name' }).getResult();

		const description = BudgetDescriptionValueObject.create('updated_description').getResult();
		budgetBox.changeDescription(description);

		expect(budgetBox.description.value).toBe('updated_description');
	});
});
