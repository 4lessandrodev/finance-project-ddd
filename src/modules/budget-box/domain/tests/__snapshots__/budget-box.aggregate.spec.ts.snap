// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget-box.aggregate should create a valid budget-box aggregate with 100% if provide not percentage 1`] = `
Object {
  "balanceAvailable": Object {
    "currency": "BRL",
    "value": 1000,
  },
  "budgetPercentage": 50,
  "createdAt": 2022-01-01T00:00:00.000Z,
  "deletedAt": undefined,
  "description": "valid_description",
  "id": "valid_id",
  "isDeleted": false,
  "isPercentage": true,
  "ownerId": "valid_owner_id",
  "reasons": Array [
    Object {
      "createdAt": 2022-01-01T00:00:00.000Z,
      "deletedAt": undefined,
      "description": "valid_description",
      "id": "valid_id",
      "isDeleted": false,
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
  ],
  "updatedAt": 2022-01-01T00:00:00.000Z,
}
`;
